There seems to be some problem with the `DeclEnumType` class that is causing this code to blow up when looping on the Builds:

```
[vagrant@localhost vagrant]$ bodhi-push --config /vagrant/development.ini
WARNING:fedmsg.crypto.x509:Crypto disabled ImportError('No module named M2Crypto',)
Traceback (most recent call last):
  File "/usr/bin/bodhi-push", line 9, in <module>
    load_entry_point('bodhi-server', 'console_scripts', 'bodhi-push')()
  File "/usr/lib/python2.7/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python2.7/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python2.7/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/vagrant/bodhi/server/push.py", line 97, in push
    for update in query.all():
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/orm/query.py", line 2613, in all
    return list(self)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/orm/query.py", line 2761, in __iter__
    return self._execute_and_instances(context)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/orm/query.py", line 2776, in _execute_and_instances
    result = conn.execute(querycontext.statement, self._params)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/base.py", line 914, in execute
    return meth(self, multiparams, params)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/sql/elements.py", line 323, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/base.py", line 1010, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/base.py", line 1078, in _execute_context
    None, None)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/base.py", line 1341, in _handle_dbapi_exception
    exc_info
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/util/compat.py", line 202, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/base.py", line 1073, in _execute_context
    context = constructor(dialect, self, conn, *args)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/default.py", line 600, in _init_compiled
    for key in compiled_params
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/default.py", line 600, in <genexpr>
    for key in compiled_params
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/sql/type_api.py", line 982, in process
    return process_param(value, dialect)
  File "/vagrant/bodhi/server/models/enum.py", line 109, in process_bind_param
    return value.value
sqlalchemy.exc.StatementError: (exceptions.AttributeError) 'unicode' object has no attribute 'value' [SQL: u'SELECT updates.id AS updates_id, updates.title AS updates_title, updates.karma AS updates_karma, updates.autokarma AS updates_autokarma, updates.stable_karma AS updates_stable_karma, updates.unstable_karma AS updates_unstable_karma, updates.requirements AS updates_requirements, updates.require_bugs AS updates_require_bugs, updates.require_testcases AS updates_require_testcases, updates.notes AS updates_notes, updates.type AS updates_type, updates.status AS updates_status, updates.request AS updates_request, updates.severity AS updates_severity, updates.suggest AS updates_suggest, updates.locked AS updates_locked, updates.pushed AS updates_pushed, updates.critpath AS updates_critpath, updates.close_bugs AS updates_close_bugs, updates.date_submitted AS updates_date_submitted, updates.date_modified AS updates_date_modified, updates.date_approved AS updates_date_approved, updates.date_pushed AS updates_date_pushed, updates.date_testing AS updates_date_testing, updates.date_stable AS updates_date_stable, updates.date_locked AS updates_date_locked, updates.alias AS updates_alias, updates.old_updateid AS updates_old_updateid, updates.release_id AS updates_release_id, updates.user_id AS updates_user_id, releases_1.id AS releases_1_id, releases_1.name AS releases_1_name, releases_1.long_name AS releases_1_long_name, releases_1.version AS releases_1_version, releases_1.id_prefix AS releases_1_id_prefix, releases_1.branch AS releases_1_branch, releases_1.dist_tag AS releases_1_dist_tag, releases_1.stable_tag AS releases_1_stable_tag, releases_1.testing_tag AS releases_1_testing_tag, releases_1.candidate_tag AS releases_1_candidate_tag, releases_1.pending_signing_tag AS releases_1_pending_signing_tag, releases_1.pending_testing_tag AS releases_1_pending_testing_tag, releases_1.pending_stable_tag AS releases_1_pending_stable_tag, releases_1.override_tag AS releases_1_override_tag, releases_1.state AS releases_1_state, comments_1.id AS comments_1_id, comments_1.karma AS comments_1_karma, comments_1.karma_critpath AS comments_1_karma_critpath, comments_1.text AS comments_1_text, comments_1.anonymous AS comments_1_anonymous, comments_1.timestamp AS comments_1_timestamp, comments_1.update_id AS comments_1_update_id, comments_1.user_id AS comments_1_user_id, releases_2.id AS releases_2_id, releases_2.name AS releases_2_name, releases_2.long_name AS releases_2_long_name, releases_2.version AS releases_2_version, releases_2.id_prefix AS releases_2_id_prefix, releases_2.branch AS releases_2_branch, releases_2.dist_tag AS releases_2_dist_tag, releases_2.stable_tag AS releases_2_stable_tag, releases_2.testing_tag AS releases_2_testing_tag, releases_2.candidate_tag AS releases_2_candidate_tag, releases_2.pending_signing_tag AS releases_2_pending_signing_tag, releases_2.pending_testing_tag AS releases_2_pending_testing_tag, releases_2.pending_stable_tag AS releases_2_pending_stable_tag, releases_2.override_tag AS releases_2_override_tag, releases_2.state AS releases_2_state, builds_1.id AS builds_1_id, builds_1.nvr AS builds_1_nvr, builds_1.epoch AS builds_1_epoch, builds_1.inherited AS builds_1_inherited, builds_1.package_id AS builds_1_package_id, builds_1.release_id AS builds_1_release_id, builds_1.update_id AS builds_1_update_id, builds_1.signed AS builds_1_signed, packages_1.id AS packages_1_id, packages_1.name AS packages_1_name, packages_1.requirements AS packages_1_requirements, packages_1.stack_id AS packages_1_stack_id, users_1.id AS users_1_id, users_1.name AS users_1_name, users_1.email AS users_1_email, users_1.show_popups AS users_1_show_popups, buildroot_overrides_1.id AS buildroot_overrides_1_id, buildroot_overrides_1.build_id AS buildroot_overrides_1_build_id, buildroot_overrides_1.submitter_id AS buildroot_overrides_1_submitter_id, buildroot_overrides_1.notes AS buildroot_overrides_1_notes, buildroot_overrides_1.submission_date AS buildroot_overrides_1_submission_date, buildroot_overrides_1.expiration_date AS buildroot_overrides_1_expiration_date, buildroot_overrides_1.expired_date AS buildroot_overrides_1_expired_date \nFROM updates LEFT OUTER JOIN releases AS releases_1 ON releases_1.id = updates.release_id LEFT OUTER JOIN comments AS comments_1 ON updates.id = comments_1.update_id LEFT OUTER JOIN builds AS builds_1 ON updates.id = builds_1.update_id LEFT OUTER JOIN releases AS releases_2 ON releases_2.id = builds_1.release_id LEFT OUTER JOIN packages AS packages_1 ON packages_1.id = builds_1.package_id LEFT OUTER JOIN buildroot_overrides AS buildroot_overrides_1 ON builds_1.id = buildroot_overrides_1.build_id LEFT OUTER JOIN users AS users_1 ON users_1.id = buildroot_overrides_1.submitter_id \nWHERE updates.request IN (%(request_1)s, %(request_2)s) ORDER BY comments_1.timestamp'] [parameters: [immutabledict({})]]
```

I'm going to try to reproduce this outside of your PR, and if I can I'll fix it in another PR.

--- a/alembic/versions/3cde3882442a_add_signed_column_to_builds.py
+++ b/alembic/versions/3cde3882442a_add_signed_column_to_builds.py
@@ -1,0 +1,26 @@
+"""Add a signed column to the Builds model.
+
+Revision ID: 3cde3882442a
+Revises: 4df1fcd59050
+Create Date: 2016-10-11 19:37:30.847965
+
+"""
+
+# revision identifiers, used by Alembic.
+revision = '3cde3882442a'
+down_revision = '4df1fcd59050'
+
+from alembic import op
+import sqlalchemy as sa
+
+
+def upgrade():
+    op.add_column('builds', sa.Column('signed', sa.Boolean(), nullable=True, default=False))
+    builds = sa.sql.table('builds', sa.sql.column('signed', sa.Boolean))
+    op.execute(
+        builds.update().values({'signed': True}))
+    op.alter_column('builds', column_name='signed', nullable=False)
+
+
+def downgrade():
+    op.drop_column('builds', 'signed')

--- a/bodhi/server/consumers/signed.py
+++ b/bodhi/server/consumers/signed.py
@@ -1,0 +1,84 @@
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+"""
+The "signed handler".
+
+This module is responsible for marking builds as "signed" when they get moved
+from the pending-signing to pending-updates-testing tag by RoboSignatory.
+"""
+
+import logging
+import pprint
+
+import fedmsg.consumers
+from pyramid.paster import get_appsettings
+from sqlalchemy import engine_from_config
+
+from bodhi.server.models import Base, Build, Release
+from bodhi.server.util import transactional_session_maker
+
+log = logging.getLogger('bodhi')
+
+
+class SignedHandler(fedmsg.consumers.FedmsgConsumer):
+    """The Bodhi Signed Handler.
+
+    A fedmsg listener waiting for messages from koji about builds being tagged.
+
+    """
+    config_key = 'signed_handler'
+
+    def __init__(self, hub, *args, **kwargs):
+        config_uri = '/etc/bodhi/production.ini'
+        self.settings = get_appsettings(config_uri)
+        engine = engine_from_config(self.settings, 'sqlalchemy.')
+        Base.metadata.create_all(engine)
+        self.db_factory = transactional_session_maker(engine)
+
+        prefix = hub.config.get('topic_prefix')
+        env = hub.config.get('environment')
+        self.topic = [
+            prefix + '.' + env + '.buildsys.tag'
+        ]
+
+        log.info('Bodhi signed handler listening on:\n'
+                 '%s' % pprint.pformat(self.topic))
+
+    def consume(self, message):
+        msg = message['body']['msg']
+        topic = message['topic']
+
+        log.info("Signed Handler handling  %s, %s" % (alias, topic))
+
+        build_nvr = '%(name)s-%(version)s-%(release)s' % msg
+        tag = msg['tag']
+
+        log.info("%s tagged into %s" % (build_nvr, tag))
+
+        with self.db_factory() as session:
+            build = Build.get(build_nvr, session)
+            if not build:
+                log.info("Build was not submitted, skipping")
+                return
+
+            if build.release.pending_testing_tag != tag:
+                log.info("Tag is not pending_testing tag, skipping")
+                return
+
+            # This build was moved into the pending_testing tag for the applicable release, which
+            # is done by RoboSignatory to indicate that the build has been correctly signed and
+            # written out. Mark it as such.
+            log.info("Build has been signed, marking")
+            build.signed = True
+            log.info("Build %s has been marked as signed" % build_nvr)

--- a/bodhi/server/models/models.py
+++ b/bodhi/server/models/models.py
@@ -487,6 +487,7 @@ class Build(Base):
     package_id = Column(Integer, ForeignKey('packages.id'))
     release_id = Column(Integer, ForeignKey('releases.id'))
     update_id = Column(Integer, ForeignKey('updates.id'))
+    signed = Column(Boolean, default=False, nullable=False)
 
     release = relationship('Release', backref='builds', lazy=False)
 
@@ -864,6 +865,10 @@ class Update(Base):
             update=up, agent=request.user.name, new_bugs=new_bugs))
 
         return up, caveats
+
+    @property
+    def signed(self):
+        return all([build.signed for build in self.builds])
 
     def obsolete_older_updates(self, db):
         """Obsolete any older pending/testing updates.

--- a/bodhi/server/push.py
+++ b/bodhi/server/push.py
@@ -16,13 +16,18 @@ The tool for triggering updates pushes.
 """
 
 import click
+from datetime import datetime
 import json
 import glob
 
 from collections import defaultdict
-from fedora.client.bodhi import Bodhi2Client
+from pyramid.paster import get_appsettings
+from sqlalchemy import engine_from_config
+from sqlalchemy.sql import or_
 
 import bodhi.server.notifications
+from bodhi.server.util import transactional_session_maker
+from bodhi.server.models import Update, Base, UpdateRequest
 
 
 @click.command()
@@ -33,18 +38,17 @@ import bodhi.server.notifications
         help='Push updates with a specific request (default: testing,stable)')
 @click.option('--builds', help='Push updates for specific builds')
 @click.option('--username', envvar='USERNAME')
-@click.option('--password', prompt=True, hide_input=True)
 @click.option('--cert-prefix', default="shell",
               help="The prefix of a fedmsg cert used to sign the message.")
+@click.option('--config', help='Configuration file to use for database credentials',
+              default='/etc/bodhi/production.ini')
 @click.option('--staging', help='Use the staging bodhi instance',
               is_flag=True, default=False)
 @click.option('--resume', help='Resume one or more previously failed pushes',
               is_flag=True, default=False)
-def push(username, password, cert_prefix, **kwargs):
+def push(username, cert_prefix, config, **kwargs):
     staging = kwargs.pop('staging')
     resume = kwargs.pop('resume')
-    client = Bodhi2Client(username=username, password=password,
-                          staging=staging)
 
     lockfiles = defaultdict(list)
     locked_updates = []
@@ -59,31 +63,51 @@ def push(username, password, cert_prefix, **kwargs):
             lockfiles[lockfile].append(update)
             locked_updates.append(update)
 
-    # If we're resuming a push
-    if resume:
-        updates = []
-        for lockfile in lockfiles:
-            doit = raw_input('Resume %s? (y/n)' % lockfile).strip().lower()
-            if doit == 'n':
-                continue
+    settings = get_appsettings(config)
+    engine = engine_from_config(settings, 'sqlalchemy.')
+    Base.metadata.create_all(engine)
+    db_factory = transactional_session_maker(engine)
 
-            for update in lockfiles[lockfile]:
-                updates.append(update)
-                click.echo(update)
-    else:
-        # release->request->updates
-        releases = defaultdict(lambda: defaultdict(list))
-        updates = []
+    update_titles = None
 
-        # Gather the list of updates based on the query parameters
-        # Since there's currently no simple way to get a list of all updates with
-        # any request, we'll take a comma/space-delimited list of them and query
-        # one at a time.
-        requests = kwargs['request'].replace(',', ' ').split(' ')
-        del(kwargs['request'])
-        for request in requests:
-            resp = client.query(request=request, **kwargs)
-            for update in resp.updates:
+    with db_factory() as session:
+        updates = []
+        # If we're resuming a push
+        if resume:
+            for lockfile in lockfiles:
+                doit = raw_input('Resume %s? (y/n)' % lockfile).strip().lower()
+                if doit == 'n':
+                    continue
+
+                for update in lockfiles[lockfile]:
+                    update = Update.get(update).first()
+                    updates.append(update)
+                    click.echo(update)
+        else:
+            # Accept both comma and space separated request list
+            requests = kwargs['request'].replace(',', ' ').split(' ')
+            requests = [UpdateRequest.from_string(val) for val in requests]
+
+            query = session.query(Update).filter(Update.request.in_(requests))
+
+            if kwargs.get('builds'):
+                query = query.join(Update.builds)
+                query = query.filter(or_(*[Build.nvr==build for build in kwargs['builds']]))
+
+            if kwargs.get('releases'):
+                releases = []
+                for r in kwargs['releases']:
+                    release = db.query(Release).filter(
+                        or_(Release.name==r,
+                            Release.name==r.upper(),
+                            Release.version==r)).first()
+                    if not release:
+                        click.echo('Unknown release: %s' % r)
+                    else:
+                        releases.append(r)
+                query = query.filter(or_(*[Update.release==r for r in releases]))
+
+            for update in query.all():
                 # Skip locked updates that are in a current push
                 if update.locked:
                     if update.title in locked_updates:
@@ -94,29 +118,31 @@ def push(username, password, cert_prefix, **kwargs):
                         click.echo('Warning: %s is locked but not in a push' %
                                    update.title)
 
-                updates.append(update.title)
-                for build in update.builds:
-                    releases[update.release.name][request].append(build.nvr)
-            while resp.page < resp.pages:
-                resp = client.query(request=request, page=resp.page + 1, **kwargs)
-                for update in resp.updates:
-                    updates.append(update.title)
-                    for build in update.builds:
-                        releases[update.release.name][request].append(build.nvr)
+                # Skip unsigned updates (this checks that all builds in the update are signed)
+                if not update.signed:
+                    click.echo('Warning: %s has unsigned builds and has been skipped' %
+                               update.title)
+                    continue
 
-            # Write out a file that releng uses to pass to sigul for signing
-            # TODO: in the future we should integrate signing into the workflow
-            for release in releases:
-                output_filename = request.title() + '-' + release
-                click.echo(output_filename + '\n==========')
-                with file(output_filename, 'w') as out:
-                    for build in releases[release][request]:
-                        out.write(build + '\n')
-                        click.echo(build)
-                click.echo('')
+                updates.append(update)
 
-    doit = raw_input('Push these %d updates? (y/n)' % len(updates)).lower().strip()
-    if doit == 'y':
+        for update in updates:
+            print update.title
+
+        doit = raw_input('Push these %d updates? (y/n)' % len(updates)).lower().strip()
+        if doit == 'y':
+            click.echo('\nLocking updates...')
+            for update in updates:
+                update.locked = True
+                update.date_locked = datetime.utcnow()
+
+            update_titles = list([update.title for update in updates])
+        else:
+            click.echo('\nAborting push')
+            raise Exception('Aborting push')
+
+
+    if update_titles:
         click.echo('\nSending masher.start fedmsg')
         # Because we're a script, we want to send to the fedmsg-relay,
         # that's why we say active=True
@@ -124,15 +150,12 @@ def push(username, password, cert_prefix, **kwargs):
         bodhi.server.notifications.publish(
             topic='masher.start',
             msg=dict(
-                updates=list(updates),
+                updates=update_titles,
                 resume=resume,
                 agent=username,
             ),
             force=True,
         )
-    else:
-        click.echo('\nAborting push')
-
 
 if __name__ == '__main__':
     push()

--- a/bodhi/server/templates/update.html
+++ b/bodhi/server/templates/update.html
@@ -374,6 +374,11 @@ $(document).ready(function(){
             <tr class="media">
               <td>${build.nvr}</td>
               <td class="pull-right">
+                % if build.signed:
+                  <span class="glyphicon glyphicon-lock" alt="Build signed"></span>
+                % else:
+                  <span class="glyphicon glyphicon-unlock" alt="Build not signed"></span>
+                % endif
                 <a href='${request.route_url("updates_rss") + "?packages=" + build.package.name}'>
                   <span class="fa fa-rss" data-toggle="tooltip" data-placement="top" title="RSS feed for new Bodhi updates containing ${build.package.name}"></span>
                 </a>
@@ -442,7 +447,11 @@ $(document).ready(function(){
 
       % if update.request:
       <tr>
-        <td>Request</td><td>${self.util.request2html(update.request) | n}</td>
+        <td>Request</td><td>${self.util.request2html(update.request) | n}
+        % if not update.signed:
+          <span class="glyphicon glyphicon-unlock" alt="Update contains unsigned builds"></span>
+        % endif
+        </td>
       </tr>
       % endif
 

--- a/fedmsg.d/signed_handler.py
+++ b/fedmsg.d/signed_handler.py
@@ -1,0 +1,3 @@
+config = dict(
+    signed_handler=True,
+)

--- a/setup.py
+++ b/setup.py
@@ -191,6 +191,7 @@ setup(name='bodhi-server',
       [moksha.consumer]
       masher = bodhi.server.consumers.masher:Masher
       updates = bodhi.server.consumers.updates:UpdatesHandler
+      signed = bodhi.server.consumers.signed:SignedHandler
       """,
       paster_plugins=['pyramid'],
       )

